### DISCUSSION TITLE
JPA에서 ID 생성 전략에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by gyungchan-jo** [2024-woowacourse-study/level-interview#221] [Spring Data]
.

**Comment by gyungchan-jo** [comment]
ID 생성 전략은 `IDENTITY`, `SEQUENCE`, `TABLE`, `AUTO` 와 hibernate6에서 추가된 `UUID까지 총 5가지가 있습니다.

1. IDENTITY
IDENTITY 전략은 엔티티가 데이터베이스에 삽입될 때 데이터베이스가 자동으로 기본 키를 생성하도록 합니다.
특징: 데이터베이스가 ID 값을 생성하므로, 엔티티가 persist될 때 ID가 할당되어 쓰기 지연 기능을 이용 못한다는 단점이 있습니다.

2. SEQUENCE
SEQUENCE 전략은 데이터베이스 시퀀스를 사용하여 ID를 생성합니다.
특징: 시퀀스를 사용하여 ID를 미리 생성할 수 있습니다. 데이터베이스에 독립적인 시퀀스 객체가 필요하고, pk 값을 가져오기 위해 DB Sequence에서 값을 조회하게 되어 이로 인해 성능상 단점이 있습니다.

3. TABLE
TABLE 전략은 별도의 테이블을 사용하여 ID를 생성합니다.
특징: 별도의 테이블에서 ID를 관리하므로 데이터베이스에 독립적입니다. SEQUENCE 전략과 비슷하게 TABLE에서 값을 조회하게 되어 성능이 떨어질 수 있습니다.

4. AUTO
AUTO 전략은 JPA가 사용 중인 데이터베이스에 따라 적절한 ID 생성 전략을 자동으로 선택합니다.
특징: 데이터베이스에 따라 다른 전략을 사용할 수 있습니다. hibernate 버전에 따라 선택되는 전략이 달라진다는 단점이 있습니다.

5. UUID
UUID 전략은 말 그대로 UUID 를 사용하여 기본 키를 생성하는 전략입니다.
특징: UUID를 저장하므로 저장 공간을 더 많이 차지하고 순차적이지 않기 때문에 인덱스의 성능이 저하된다는 단점이 있습니다.

**Comment by nayonsoso** [comment]
### 직접 할당
- `@Id` 어노테이션만 사용하여 Id값을 직접 할당할 수 있다.

### 자동 할당
- `@Id`와 `@GeneratiedValue`를 함께 사용해서 원하는 키 생성 전략을 선택할 수 있다.
- `@GeneratiedValue`의 stretagy 옵션을 통해 설정할 수 있는데, 여기에 올 수 있는 값인 GenerationType는 다음과 같다.
```java
public enum GenerationType { 
    IDENTITY,
    SEQUENCE, 
    UUID,
    TABLE, 
    AUTO
}
```

#### IDENTITY
- 기본 키 생성을 DB에 위임하는 전략이다.
- 이 전략을 사용하면 엔티티를 생성할 때 쓰기지연이 적용되지 않는다.
  - 왜냐하면 jpa 에서 어떤 엔티티를 영속하기 위해선 식별자가 필요한데, IDENTITY 전략에서는 이 식별자가 DB에 저장되어야 할당되기 때문이다.
  - 따라서 엔티티를 생성할 때 즉시 INSERT SQL이 실행되어야 한다.

#### SEQUENCE
- 시퀀스키 생성 전략을 지원하는 DB에서 사용할 수 있다.
- 데이터베이스 시퀀스란, 유일한 값을 자동으로 생성하게 하는 객체이다.
- auto_increment와 달리 초기 값과 한번에 증가할 크기를 설정할 수 있다.
- 이 시퀀스를 키 생성 전략으로 갖는 DB에 대해 SEQUENCE 전략을 사용할 수 있다.
- 어떤 시퀀스를 사용할 것인지를 `@SequenceGenerator` 로 설정할 수 있다.

#### TABLE
- 키 생성 전용 테이블을 만들어 시퀀스를 흉내내는 전략이다.
- 어떤 테이블을 사용할 것인지를 `@TableGenerator`로 설정할 수 있다.
- 내부 동작 방식은 SEQUENCE 방식과 동일하다.
  - DB와 한번 더 통신한다는 점에서 SEQUENCE보다 성능이 안좋다는 단점이 있지만, 
  - 모든 DB에 적용할 수 있다는 장점이 있다.

#### AUTO
- DBSM 과 기본키 자료형에 따라 자동으로 전략을 선택한다.

![](https://velog.velcdn.com/images/skyepodium/post/eac9c592-6299-49a4-94c1-5541103c8b37/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA%202022-01-02%20%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE%207.25.11.png)

출처 : https://velog.io/@skyepodium/JPA-기본키-매핑-전략
